app/state.go

```diff
@@ -808,7 +808,9 @@ func (u *updateStoreState) Handle(ctx *StateContext, c Controller) (State, bool)
 	installer, err := c.ReadArtifactHeaders(u.imagein)
 	if err != nil {
 		log.Errorf("Fetching Artifact headers failed: %s", err)
-		return NewUpdateStatusReportState(&u.update, client.StatusFailure), false
+		merr := NewTransientError(fmt.Errorf("Error reading artifact headers: %s", err.Error()))
+
+		return NewUpdateErrorState(merr, &u.update), false
```

Comment:
Hmm, I think the `UpdateErrorState` leads to `ArtifactFailure` being executed. However, we cannot execute any `Artifact` states unless `Download` has completed to successfully (for several reasons, but most importantly because it's insecure).

So I think this should be `NewUpdateCleanupState` instead. I think it should still execute error scripts as long as [the correct status](https://github.com/mendersoftware/mender/blob/3.5.x/app/state.go#L1388-L1392) is passed into it.

Other than that I agree with the fix!

Reply:
Okay, I see, my thinking was that UpdateErrorState{} should trigger the execution of the error script in transitionState(), but I now see that going into NewUpdateCleanUpdateState{} is a better and more secure solution.Thanks :)